internal/liqonet/tunnel-operator/reflected_endpointslice_controller.go

```diff
@@ -0,0 +1,283 @@
+
+				// Soft delete object
+				endpointsInfo := value.(map[string]liqoiptables.EndpointInfo)
+				for endpoint, endpointInfo := range endpointsInfo {
+					endpointInfo.Deleting = true
+					endpointsInfo[endpoint] = endpointInfo
+				}
+				r.endpointslicesInfo.Store(nsName, endpointsInfo)
+
+				if err := r.gatewayNetns.Do(r.ensureIptablesRules); err != nil {
+					klog.Errorf("Error while ensuring iptables rules: %w", err)
+					return ctrl.Result{}, err
+				}
+
+				// Hard delete object
+				r.endpointslicesInfo.Delete(nsName)
+			}
+		}
+		return ctrl.Result{}, err
+	}
+
+	clusterSelector := nsOffloading.Spec.ClusterSelector
+
+	nodes := corev1.NodeList{}
+	if err := r.List(ctx, &nodes); err != nil {
+		klog.Error(err)
+		return ctrl.Result{}, err
+	}
+
+	// Build endpointInfo objects
+	endpointsInfo := map[string]liqoiptables.EndpointInfo{}
+	//For each endpoint, find ClusterIDs of clusters that can reach that endpoint
+	for _, endpoint := range endpointslice.Endpoints {
+		clusterIDs := []string{}
+
+		for _, node := range nodes.Items {
+
+			if *endpoint.NodeName == node.Name {
+				continue
+			}
+
+			matchClusterSelctor, err := nsoffctrl.MatchNodeSelectorTerms(&node, &clusterSelector)
+			if err != nil {
+				klog.Error(err)
+				return ctrl.Result{}, err
+			}
+
+			if matchClusterSelctor {
+				if clusterID, virtual := utils.GetNodeClusterID(&node); virtual {
+					clusterIDs = append(clusterIDs, clusterID)
+				}
+			}
+		}
+
+		endpointsInfo[endpoint.Addresses[0]] = liqoiptables.EndpointInfo{Address: endpoint.Addresses[0], SrcClusterIDs: clusterIDs}
+	}
+
+	// Check if the object is under deletion
+	if !endpointslice.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Endpointslice under deletion: skip creation of iptables rules and return no error
+		klog.Infof("Endpointslice %q under deletion: skipping iptables rules update", nsName)
+		return ctrl.Result{}, nil
+	}
+
+	//Check if endpoint(s) are no more part of the endpointslice
+	value, loaded := r.endpointslicesInfo.Load(nsName)
+	if loaded {
+		oldEndpointsInfo := value.(map[string]liqoiptables.EndpointInfo)
+		for oldEndpoint, oldEndpointInfo := range oldEndpointsInfo {
+			if _, ok := endpointsInfo[oldEndpoint]; !ok {
+				oldEndpointInfo.Deleting = true
+				endpointsInfo[oldEndpoint] = oldEndpointInfo
+			}
+		}
+	}
+
+	// Check if there aren't new information: in this case it's not necessary ensure iptables rules
+	if len(endpointsInfo) == 0 {
+		// Endpoints fiels of Endpoinslice yet empty: skip creation of iptables rules and return no error
+		klog.Infof("Endpoints of endpointslice %q not yet set: skipping iptables rules update", nsName)
+		return ctrl.Result{}, nil
+	}
+
+	// Store endpointslicesInfo object
+	r.endpointslicesInfo.Store(nsName, endpointsInfo)
+
+	// Ensure iptables rules
+	klog.Infof("Ensuring updated iptables rules")
+	if err := r.gatewayNetns.Do(r.ensureIptablesRules); err != nil {
+		klog.Errorf("Error while ensuring iptables rules: %w", err)
```

Comment:
You don't need to log it, just create a new error with fmt.Errorf and return it.